decision-records/2024-06-04--service-to-stop-while-waiting-for-cert.md

```diff
@@ -0,0 +1,106 @@
+**Date:** 2024-06-04<br/>
+**Author:** @sed-i
+
+## Context and Problem Statement
+When an admin forms a tls relation, there is a time window in which:
+
+1. the juju model tells us we must operate with TLS enabled; however
+2. we haven't received the server cert yet.
+
+As a result we had some code ordering issues (e.g. #414):
+- Should we update the config file with the tls config, which includes path to
+  the cert, before we got the cert, only to perfectly reflect the juju model?
+- Should we wait with the tls config and allow the workload to continue to
+  operate in plain http, in contradiction to the juju model?
+- How should we determine within the charm if "TLS is enabled"?
+
+This problem is common to many charms.
+
+## Decision
+A server shall NOT keep using plain http while waiting for a cert. The service
+must be stopped until the cert is in place.
+The charm shall have a "Waiting" status during that period (subject to other
+status priorities).
+
+We also distinguish between several substates:
+- TLS disabled: a tls-certificate is not in place, no certs on disk, no tls
+  section in app config, app is running in plain HTTP mode.
+- TLS enabled: a tls-certificate relation is present.
+- Cert ready: we have a cert in relation data.
+- TLS config ready: the cert that is in relation data was written to disk, and
+  app config has a corresponding tls section.
+- TLS active: the app is running with the new config.
+
+```mermaid
+stateDiagram-v2
+    [*] --> TLSDisabled : install
+
+    TLSDisabled: TLS disabled
+    note right of TLSDisabled
+        No TLS relation
+        No certs on disk
+        No TLS section in app config
+        App running in plain HTTP mode
+    end note
+    TLSDisabled --> TLSEnabled : tls-relation-joined
+
+    OUT: Service outage
+    state OUT {
+        TLSEnabled: TLS enabled
+        note left of TLSEnabled
+            TLS relation present
+            CSR sent, waiting for cert, or cert expired or revoked
+        end note
+        TLSEnabled --> CertReady: tls-relation-changed
+        CertReady: Cert ready
+        note left of CertReady
+            A live cert is in relation data
+        end note
+        CertReady --> TLSReady
+        TLSReady: TLS config ready
+        note right of TLSReady
+            The cert that is in relation data was written to disk
+            App config has a corresponding tls section
+        end note
+        TLSReady --> TLSActive
+    }
+
+    TLSActive: TLS active
+    note right of TLSActive
+        App restarted and operates in TLS
+    end note
+    TLSActive --> TLSActive : cert update
+
+    TLSActive --> TLSEnabled : cert revoked, cert expired
+```
+
+## Benefits
+- No need to worry if it takes the cert a long time to arrive.
```

Comment:
or if the charm remains in any other inconsistent transient state for 'long'

Reply:
The other inconsistent states are not a new concern so I think it's out of scope.
I.e. if pebble push or update cert fails is not specific to this ADR.